# Computational G4 analysis

This is a collection of scripts that have been used to carry out the bioinformatic G4 prediction and analysis for Hile et al. "Replicative DNA Polymerase Epsilon and Delta Holoenzymes Show Wide Ranging Inhibition at G-Quadruplexes in the Human Genome". The overall workflow is as follows: 1) *In silico* prediction of G4 motifs from the human CHM13v2.0 reference assembly with *Quadron*, 2) Finding most abundant G4 motifs 3) Find overlaps with repetitive elements 4) Determining "exact" vs "embedded" match. 4) Finding overlaps with genes and promoters

## *In silico* prediction of G4 motifs

To computationally predict G4 motifs, we used *quadron* with default parameters.

```r
################################################################################                                                                                              
# Requires the libraries "doMC", "foreach", "itertools", "xgboost" (0.4-4),    #                                                                                              
# "caret" and "plyr".                                                          #                                                                                              
# If not already installed in R, you can install those by typing:              #                                                                                              
# install.packages(c("doMC", "foreach", "itertools", "plyr", "caret"))         #                                                                                              
# Specific steps are needed to install the xgboost version 0.4-4, as detailed  #                                                                                              
# in the Readme file.                                                          #                                                                                              
# The default fastread==TRUE option in readfasta requires "data.table" library.#                                                                                              
################################################################################                                                                                              
#setwd("./lib")                                                                                                                                                               
#source("bitcompile.R")                                                                                                                                                       
#rm(list=ls())                                                                                                                                                                
#setwd("..")                                                                                                                                                                  
.libPaths(c(.libPaths(), "/opt/bx/R/xgboost_0.4-4"))                                                                                                                          

print("NOTE: Loading Quadron core...", quote=FALSE)                                                                                                                           
load("[local_path]/Quadron-master/Quadron.lib")                                                                                                              


args = commandArgs(trailingOnly=TRUE)                                                                                                                                         
Quadron(FastaFile= paste0("[local_path]/CHM13v.0/single_chromosomes/", args[1], ".fa"),                                                
        OutFile  = paste0("[local_path]/quadron/CHM13v2.0_output/",args[1], ".txt"),                                                       
        nCPU     = 8,                                                                                                                                                         
        SeqPartitionBy = 1000000)                                                                                                                                             

#file.remove("Quadron.lib")                                                                                                                                                   
#rm(list=ls())                                                                                                                                                                
################################################################################
```

To convert the *Quadron* output to a bed file, the one bp needs to be subtracted from the start and end coordinates (2nd and 3rd column in the output).

Most abundant motifs were identified with `cut -f 6 Quadron_output.bed | sort | uniq -c | sort -k 1 -n`. The 6th column contains the nucleotide sequence of the G4 motif, therefore the sort and uniq commands rank them by abundance.  The most abundant motifs were then intersected with repetitive elements using `bedtools intersect `with the CHM13v2.0 repeat masker annotation (`chm13v2.0_RepeatMasker_4.1.2p1.2022Apr14.bed`) , leading to the identification of the SVA, L1, GGGA, etc. motifs and their reverse complements.

## "Exact match" vs. "match" of motifs

The "GGGAGGGAGGTGGGGGGG" motif is the most abundant motif in CHM13v2.0 and present predominantly in the SVA_F element, but it also occurs embedded in other motifs that were identified by *Quadron* as potentially G4-forming. In the simples case, another nucleotide was added to the sequence:
Exact match: **GGGAGGGAGGTGGGGGGG** vs. match: **GGGAGGGAGGTGGGGGGG**G, or another example for an exact match **GGGGACTGTTGTGGGGTGGGGGGAGGGGGGAGGG** vs.  **GGGGACTGTTGTGGGGTGGGGGGAGGGGGGAGGG**ATAGCATTGGG. 
To determine the number of exact v.s. embedded, we ran the following `bash` script:

    #!/bin/bash
    while IFS= read -r line; do                                                                                                                                                                                                                                                      
            id=$(echo $line | awk '{print $1}')                                                                                                                                                                                                                                      
            motif=$(echo $line | awk '{print $2}')                                                                                                                                                                                                                                   
            matches=$(awk -v m=$motif '$6!=m' Quadron_chm13v2.0.bed | grep $motif | wc | awk '{print $1}')                                                                                                                                                                           
            exact_matches=$(awk -v m=$motif '$6==m' Quadron_chm13v2.0.bed | wc | awk '{print $1}')                                                                                                                                                                                   
            quadron_score_exact_matches=$(awk -v m=$motif '$6==m' Quadron_chm13v2.0.bed | awk '{ sum += $5; n++ } END { if (n > 0) print sum / n; }')                                                                                                                                
            quadron_score_matches=$(awk -v m=$motif '$6!=m' Quadron_chm13v2.0.bed | grep $motif |  awk '{ sum += $5; n++ } END { if (n > 0) print sum / n; }')                                                                                                                       
            echo -e $id '\t'$motif '\t'  $matches '\t' $quadron_score_matches '\t' $exact_matches '\t' $quadron_score_exact_matches                                                                                                                                                  
            awk -v m=$motif '$6!=m' Quadron_chm13v2.0.bed | grep $motif | awk -v m=$motif -v i=$id '{print $0 "\t" m "\t" i "\t" "match"}' >> output                                                                                                                               
            awk -v m=$motif '$6==m' Quadron_chm13v2.0.bed | awk -v m=$motif -v i=$id '{print $0 "\t" m "\t" i "\t" "exact_match"}' >> output                                                                                                                                       
                                                                                                                                                                                                                                                                                     
    done<most_abundant_motifs 

## Overlaps with genes and promoters

Finally, we wanted to investigate whether the motifs we've identified as the most abundant also overlap genic and promoter regions. For this we used the genome annotation liftover from hg38 to CHM13v2.0 (`Homo_sapiens-GCA_009914755.4-2022_07-genes.gff3`) and defined promoters broadly as the 1,000-bp region upstream of the start of a gene. We then used `bedtools` to intersect the motif coordinates with the gene annotation and promoters:

```bash
#!/bin/bash                                                                                                                                                                                                                                                                      
                                                                                                                                                                                                                                                                                 
promoter_length=1000                                                                                                                                                                                                                                                             
                                                                                                                                                                                                                                                                                 
awk '$4=="+"' chm13v2.0.genes.bed > plus_strand_genes.bed                                                                                                                                                                                                                        
awk '$4=="-"' chm13v2.0.genes.bed > minus_strand_genes.bed                                                                                                                                                                                                                       
                                                                                                                                                                                                                                                                                 
bedtools flank -i plus_strand_genes.bed -g chm13v2.0.genome -l $promoter_length -r 0 > promoters_plus_strand.bed                                                                                                                                                                 
bedtools flank -i minus_strand_genes.bed -g chm13v2.0.genome -l 0 -r $promoter_length > promoters_minus_strand.bed                                                                                                                                                               
                                                                                                                                                                                                                                                                                 
cat promoters_plus_strand.bed promoters_minus_strand.bed | bedtools sort -i - > chm13v2.0.promoters_${promoter_length}bp.bed                                                                                                                                                     
                                                                                                                                                                                                                                                                                 
                                                                                                                                                                                                                                                                                 
while IFS= read -r line; do                                                                                                                                                                                                                                                      
exact_matches=$(awk -v m=$line '$8==m' chm13v2.0_quadron_exact_matches_corr_cordinates.bed | \                                                                                                                                                                                   
bedtools sort -i - |\                                                                                                                                                                                                                                                            
bedtools intersect -a chm13v2.0.promoters_${promoter_length}bp.bed -b - | \                                                                                                                                                                                                      
wc | awk '{print $1}')                                                                                                                                                                                                                                                           
matches=$(awk -v m=$line '$8==m' chm13v2.0_quadron_matches_corr_cordinates.bed | \                                                                                                                                                                                               
bedtools sort -i - |\                                                                                                                                                                                                                                                            
bedtools intersect -a chm13v2.0.promoters_${promoter_length}bp.bed -b - | \                                                                                                                                                                                                      
wc | awk '{print $1}')                                                                                                                                                                                                                                                           
echo -e $line '\t' $(($exact_matches+$matches))                                                                                                                                                                                                                                  
done<exact_matches_motif_list  

#while IFS= read -r line; do                                                                                                                                                                                                                                                     
#echo $line                                                                                                                                                                                                                                                                      
#gene_e=$(awk -v m=$line '$8==m' chm13v2.0_quadron_exact_matches_corr_cordinates.bed |\                                                                                                                                                                                          
#bedtools intersect -a Homo_sapiens-GCA_009914755.4-2022_07-genes.gff3  -b - | \                                                                                                                                                                                                 
#cut -f 3 | sort | uniq -c | grep 'gene' | awk '{print $1}')                                                                                                                                                                                                                     
gene_m=$(awk -v m=$line '$8==m' chm13v2.0_quadron_matches_corr_cordinates.bed |\                                                                                                                                                                                                 
bedtools intersect -a Homo_sapiens-GCA_009914755.4-2022_07-genes.gff3  -b - | \                                                                                                                                                                                                  
cut -f 3 | sort | uniq -c | grep 'gene' | awk '{print $1}')                                                                                                                                                                                                                      
                                                                                                                                                                                                                                                                                 
exon_e=$(awk -v m=$line '$8==m' chm13v2.0_quadron_exact_matches_corr_cordinates.bed |\                                                                                                                                                                                           
bedtools intersect -a Homo_sapiens-GCA_009914755.4-2022_07-genes.gff3  -b - | \                                                                                                                                                                                                  
cut -f 3 | sort | uniq -c | grep 'exon' | awk '{print $1}')                                                                                                                                                                                                                      
exon_m=$(awk -v m=$line '$8==m' chm13v2.0_quadron_matches_corr_cordinates.bed |\                                                                                                                                                                                                 
bedtools intersect -a Homo_sapiens-GCA_009914755.4-2022_07-genes.gff3  -b - | \                                                                                                                                                                                                  
cut -f 3 | sort | uniq -c | grep 'exon' | awk '{print $1}')                                                                                                                                                                                                                      
                                                                                                                                                                                                                                                                                 
intron_e=$(awk -v m=$line '$8==m' chm13v2.0_quadron_exact_matches_corr_cordinates.bed |\                                                                                                                                                                                         
bedtools intersect -a Homo_sapiens-GCA_009914755.4-2022_07-genes.gff3  -b - | \                                                                                                                                                                                                  
cut -f 3 | sort | uniq -c | grep 'intron' | awk '{print $1}')                                                                                                                                                                                                                    
intron_m=$(awk -v m=$line '$8==m' chm13v2.0_quadron_matches_corr_cordinates.bed |\                                                                                                                                                                                               
bedtools intersect -a Homo_sapiens-GCA_009914755.4-2022_07-genes.gff3  -b - | \                                                                                                                                                                                                  
cut -f 3 | sort | uniq -c | grep 'intron' | awk '{print $1}')                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                                                 
CDS_e=$(awk -v m=$line '$8==m' chm13v2.0_quadron_exact_matches_corr_cordinates.bed |\                                                                                                                                                                                            
bedtools intersect -a Homo_sapiens-GCA_009914755.4-2022_07-genes.gff3  -b - | \                                                                                                                                                                                                  
cut -f 3 | sort | uniq -c | grep 'CDS' | awk '{print $1}')                                                                                                                                                                                                                       
CDS_m=$(awk -v m=$line '$8==m' chm13v2.0_quadron_matches_corr_cordinates.bed |\                                                                                                                                                                                                  
bedtools intersect -a Homo_sapiens-GCA_009914755.4-2022_07-genes.gff3  -b - | \                                                                                                                                                                                                  
cut -f 3 | sort | uniq -c | grep 'CDS' | awk '{print $1}')                                                                                                                                                                                                                       
                                                                                                                                                                                                                                                                                 
                                                                                                                                                                                                                                                                                 
#echo -e $(($gene_e + $gene_m)) '\t' $(($exon_e + $exon_m)) '\t' $(($intron_e + $intron_m)) '\t' $(($CDS_e + $CDS_m))                                                                                                                                                            
echo -e $gene_e '\t' $gene_m '\t' $exon_e '\t' $exon_m '\t' $intron_e '\t' $intron_m '\t' $CDS_e '\t' $CDS_m                                                                                                                                                                     
done<exact_matches_motif_list

while IFS= read -r line; do                                                                                                                                                                                                                                                      
echo $line                                                                                                                                                                                                                                                                       
awk -v m=$line '$8==m' chm13v2.0_quadron_exact_matches_corr_cordinates.bed |\                                                                                                                                                                                                    
sed 's/chr//g' |\                                                                                                                                                                                                                                                                
bedtools intersect -a Homo_sapiens-GCA_009914755.4-2022_07-genes.gff3  -b - | \                                                                                                                                                                                                  
cut -f 3 | sort | uniq -c | grep 'gene' | awk '{print $1}'                                                                                                                                                                                                                       
awk -v m=$line '$8==m' chm13v2.0_quadron_matches_corr_cordinates.bed |\                                                                                                                                                                                                          
sed 's/chr//g' |\                                                                                                                                                                                                                                                                
bedtools intersect -a Homo_sapiens-GCA_009914755.4-2022_07-genes.gff3  -b - | \                                                                                                                                                                                                  
cut -f 3 | sort | uniq -c | grep 'gene' | awk '{print $1}'                                                                                                                                                                                                                       
                                                                                                                                                                                                                                                                                 
awk -v m=$line '$8==m' chm13v2.0_quadron_exact_matches_corr_cordinates.bed |\                                                                                                                                                                                                    
sed 's/chr//g' |\                                                                                                                                                                                                                                                                
bedtools intersect -a Homo_sapiens-GCA_009914755.4-2022_07-genes.gff3  -b - | \                                                                                                                                                                                                  
cut -f 3 | sort | uniq -c | grep 'exon' | awk '{print $1}'                                                                                                                                                                                                                       
awk -v m=$line '$8==m' chm13v2.0_quadron_matches_corr_cordinates.bed |\                                                                                                                                                                                                          
sed 's/chr//g' |\                                                                                                                                                                                                                                                                
bedtools intersect -a Homo_sapiens-GCA_009914755.4-2022_07-genes.gff3  -b - | \                                                                                                                                                                                                  
cut -f 3 | sort | uniq -c | grep 'exon' | awk '{print $1}'                                                                                                                                                                                                                       
                                                                                                                                                                                                                                                                                 
awk -v m=$line '$8==m' chm13v2.0_quadron_exact_matches_corr_cordinates.bed |\                                                                                                                                                                                                    
sed 's/chr//g' |\                                                                                                                                                                                                                                                                
bedtools intersect -a Homo_sapiens-GCA_009914755.4-2022_07-genes.gff3  -b - | \                                                                                                                                                                                                  
cut -f 3 | sort | uniq -c | grep 'intron' | awk '{print $1}'                                                                                                                                                                                                                     
awk -v m=$line '$8==m' chm13v2.0_quadron_matches_corr_cordinates.bed |\                                                                                                                                                                                                          
sed 's/chr//g' |\                                                                                                                                                                                                                                                                
bedtools intersect -a Homo_sapiens-GCA_009914755.4-2022_07-genes.gff3  -b - | \                                                                                                                                                                                                  
cut -f 3 | sort | uniq -c | grep 'intron' | awk '{print $1}'                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                                 
awk -v m=$line '$8==m' chm13v2.0_quadron_exact_matches_corr_cordinates.bed |\                                                                                                                                                                                                    
sed 's/chr//g' |\                                                                                                                                                                                                                                                                
bedtools intersect -a Homo_sapiens-GCA_009914755.4-2022_07-genes.gff3  -b - | \                                                                                                                                                                                                  
cut -f 3 | sort | uniq -c | grep 'CDS' | awk '{print $1}'                                                                                                                                                                                                                        
awk -v m=$line '$8==m' chm13v2.0_quadron_matches_corr_cordinates.bed |\                                                                                                                                                                                                          
sed 's/chr//g' |\                                                                                                                                                                                                                                                                
bedtools intersect -a Homo_sapiens-GCA_009914755.4-2022_07-genes.gff3  -b - | \                                                                                                                                                                                                  
cut -f 3 | sort | uniq -c | grep 'CDS' | awk '{print $1}'                                                                                                                                                                                                                        
                                                                                                                                                                                                                                                                                 
                                                                                                                                                                                                                                                                                 
echo -e $(($gene_e + $gene_m)) '\t' $(($exon_e + $exon_m)) '\t' $(($intron_e + $intron_m)) '\t' $(($CDS_e + $CDS_m))                                                                                                                                                             
echo -e $gene_e '\t' $gene_m '\t' $exon_e '\t' $exon_m '\t' $intron_e '\t' $intron_m '\t' $CDS_e '\t' $CDS_m                                                                                                                                                                     
done<exact_matches_motif_list          
```


